src/Core/System/DependencyInjection/configuration.xml

```diff
@@ -20,6 +20,13 @@
             <argument type="service" id="app.repository"/>
         </service>
 
+        <service id="Shopware\Core\System\SystemConfig\Subscriber\MemoizedSystemConfigLoaderResetSubscriber">
+            <argument type="service" id="Shopware\Core\System\SystemConfig\SystemConfigLoader"/>
```

Comment:
can't you directly inject the `MemoizedSystemConfigLoader` here and get rid of the decoration handling in the service itself?

I'm not sure how symfony handles it internally if you reference a service directly that decorates another service :thinking: 

but maybe you could test what happens if you add another decorator with a lower priority then the memoizationLoader (so the memoizationLoader is not the outermost decorator) 

to ensure that the behaviours stays the same you could also add a empty decorator in the test env, by specifiying that decorator in the `services_test.xml`, naybe that's already a good idea for the current implementation to always ensure that the service can still be decorated by others

Reply:
Not really, a decorated service replaces the original service completely. The original service is available as `<new-service-id>.inner` but that doesn't really help as I don't know the new service ID. It might be possible to do something with compiler passes but it is more complicated.

There is one alternative solution though, I could add one more service that `stores` the config but is outside of the decoration chain and use that one to store / retrieve the config in the decortor. That way I can be sure that is the correct service.

Reply:
That sounds better to me then the current state.

So we have something like a `MemoizedSystemConfigStore` that holds the `$configs` variable and knows when to clear it (so no need for a seperate event subscriber for that)

The decorator then uses the store, so in the end we would end up with the same number of services as in the current state, but by changing the service which `stores` the config we are able do get rid of the special decoration handling.

but please mark the `store` service as `@internal` as that service should not be decorated by third parties

@OliverSkroblin what do you think about that?

Reply:
@OliverSkroblin @keulinho I have now created a proposed alternative solution. Personally, I also think it is the better solution: https://github.com/shopware/platform/pull/2342/commits/53a71ea9ee9101af59c0f28596b409ca3fb87632